### PR TITLE
adds new_session_client method

### PR DESCRIPTION
**Pull request by marrony**
closes https://github.com/faunadb/sales-engineering/issues/353

### CODE REVIEW

File: faunadb/client.py [hunk 1]
```diff
@@ -89,6 +119,29 @@ def ping(self, scope=None, timeout=None):
...
+    Create a new client from the existing config with a given secret.
+    The returned client share its parent underlying resources.
+
+    :param secret:
+      Credentials to use when sending requests.
+    :param observer:
+      Callback that will be passed a :any:`RequestResult` after every completed request.
+    :return:
+    """
+    if self.counter.get() > 0:
```
Comment (erickpintor) [inline faunadb/client.py]: You should use `increment` here. There is a chance that `get` returns more than zero but immediately after some other thread decrements and close the session you're sharing with the new client.

**Comment by marrony** [comment]
do we have a verdict here on what should be the name? `new_session_client` or `with_secret`? 

FYI: my last commit made the clients share the same `Session` object with a ref counter. About add an option to increase/decrease the threadpool, the only way I see was to add a new adapter, pretty much like this example https://github.com/kennethreitz/requests/blob/master/docs/user/advanced.rst#example-specific-ssl-version

**Comment by erickpintor** [comment]
If you return a new client, I think `new_session_client` is a proper name. If you receive a lambda and just execute it on a contained method, `with_secret` seems better.

For the pool size, I think you can try to [`mount`](https://github.com/kennethreitz/requests/blob/master/requests/sessions.py#L359) http and https to the same [adapter](https://github.com/kennethreitz/requests/blob/master/requests/adapters.py#L107) but, with a different pool size.

File: faunadb/client.py [hunk 2]
```diff
@@ -1,14 +1,40 @@
...
 
+class _Counter(object):
+  def __init__(self, init_value=0):
+    self.lock = threading.Lock()
+    self.counter = init_value
+
+  def __str__(self):
+    return "Counter(%s)" % self.counter
+
+  def increment(self):
```
Comment (erickpintor) [inline faunadb/client.py]: Method not used anymore.

File: faunadb/client.py [hunk 3]
```diff
@@ -1,14 +1,40 @@
...
+    with self.lock:
+      self.counter += 1
+      return self.counter
+
+  def get_and_increment(self):
```
Comment (erickpintor) [inline faunadb/client.py]: There is still a race condition in here:

- Thread 1 calls `decrement`: decrements counter to 0 and close the session
- Thread 2 calls `get_and_increment`: increments the counter to 1, returns 0, therefore fail to create a new session client
- Thread 3 calls `get_and_increment`: increments the counter to 2, returns 1, allowing a new session client to be create with a closed session.

The knowledge of a live session must be composed by its ref count plus a flag that tells you if the session was closed already.

Comment (marrony) [inline faunadb/client.py]: I understand you reasoning but, if `decrement` returns 0 means that no other client is holding this session anymore, also, at this point, if `__del__` gets called means that this client is being garbage collected, hence, no client could possibly call `new_session_client`.

Comment (erickpintor) [inline faunadb/client.py]: You're right. Thread 1 would not decrement to zero if thread 2 and 3 are holding a client. Ignore me. :)

File: faunadb/client.py [hunk 4]
```diff
@@ -42,36 +68,49 @@ def __init__(
...
+      self.counter = kwargs['counter']
+
   def __del__(self):
     # pylint: disable=bare-except
-    try:
-      self.session.close()
-    except:
-      pass
+    if self.counter.decrement() == 0:
+      try:
```
Comment (ashfire908) [inline faunadb/client.py]: This try/except block was to attempt to handle some weird corner that came up in testing (I think it was when a test failed?), it's not a proper fix, and should be removed if possible.

Comment (marrony) [inline faunadb/client.py]: yeah, I really wanted to remove this but I was afraid to do, though, I removed and no exception was thrown

**Comment by marrony** [comment]
Can I have another review on this PR?

File: faunadb/client.py [hunk 5]
```diff
@@ -42,36 +68,45 @@ def __init__(
...
-    if secret is not None:
-      self.session.auth = (secret, "")
-
-    self.session.headers.update({
-      "Accept-Encoding": "gzip",
-      "Content-Type": "application/json;charset=utf-8"
-    })
-    self.session.timeout = timeout
-
+    self.auth = HTTPBasicAuth(secret, "")
```
Comment (erickpintor) [inline faunadb/client.py]: Just to confirm: is this base64 encrypted?

Comment (marrony) [inline faunadb/client.py]: yes

**Pull request merged**